sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java

```diff
@@ -1100,9 +1095,47 @@ public Param param(String key) {
     /**
      * Deprecated rules keys for this rule.
      * <p>
-     * Used by SonarQube to provide support for rule key renaming and/or change of repository.
+     * If you want to rename the key of a rule or change its repository or both, register the rule's previous repository
+     * and key.
+     * <br>
+     * This will allow SonarQube to support "issue renaming" (ie. <strong>not</strong> closing existing issues
+     * for this rules, created under the rule's previous repository and/or name, and recreate them as new under the
+     * new issue's repository and/or key).
+     * <p>
+     * Several deprecated keys can be provided to allow SonarQube to support "issue renaming" across multiple versions
+     * of a plugin.
+     * <br>
+     * Consider the following use case scenario:
+     * <ul>
+     *   <li>Rule {@code Foo:A} is defined in version 1 of the plugin</li>
+     *   <li>Rule's key is renamed to B in version 2 of the plugin
+     *      <ul><li>rule {@code Foo:B} defines {@code Foo:A} as a deprecated key</li></ul>
+     *   </li>
+     *   <li>All rules, including {@code Foo:B}, are moved to a new repository Bar in version 3 of the plugin
+     *      <ul><li>rule {@code Bar:B} defines {@code Foo:A} and {@code Foo:B} as deprecated keys (in order,
+     *          oldest key first, for documentation purpose).</li></ul>
+     *   </li>
+     * </ul>
+     *
+     * With all deprecated keys defined in version 3 of the plugin, SonarQube will be able to support "issue renaming"
+     * for this rule in all cases:
+     * <ul>
+     *   <li>plugin upgrade from v1 to v2,</li>
+     *   <li>plugin upgrade from v2 to v3</li>
+     *   <li>AND plugin upgrade from v1 to v3</li>
```

Comment:
@sns-seb should we mention downgrading as well?

Reply:
I will add it when we implement it, it's a "cutable" branch for this sprint :)